torch/csrc/jit/frontend/concrete_module_type.cpp

```diff
@@ -269,9 +297,10 @@ void ConcreteModuleTypeBuilder::addBuiltinFunction(
 
 void ConcreteModuleTypeBuilder::addModule(
     std::string name,
-    std::shared_ptr<ConcreteModuleType> meta) {
-  modules_.emplace_back(
-      ConcreteModuleTypeBuilder::ModuleInfo{std::move(name), std::move(meta)});
+    std::shared_ptr<ConcreteModuleType> meta,
+    TypePtr containedTypeHint) {
```

Comment:
Theres `setContainedTypeHint` which sets the type hint for all contained modules, and now we have an argument `containedTypeHint` which is only for one specific module. Maybe rename `containedTypeHint` variable name to be more specific to the module? At first glance i thought the variable was being used for the same purposed as `setContainedTypeHint`, but it's not

Reply:
Hmm, yeah, there are two sort of overlapping concepts here:
1. `ConcreteModuleTypeBuilder::containedTypeHint_`: This is a type hint for all contained modules. The `ConcreteModuleTypeBuilder` should have its `iterableModuleKind_` set to `IterableModuleKind::DICT` (but I think we can handle `IterableModuleKind::LIST` easily as well) for this to have meaning. This is the type hint that gets set and used when `self` is annotated:
```
        # Test annotation of self.
        class ModDict(torch.nn.ModuleDict):
            # Calls setContainedType on the ConcreteModuleTypeBuilder associated with ModDict
            __annotations__ = {"self": Dict[str, ModuleInterface]}

            def __init__(self):
                super().__init__({"module": ImplementsInterface()})

            def forward(self, x: torch.Tensor, key: str) -> Any:
                return self[key].forward(x)
```
2. `ConcreteModuleTypeBuilder::ModuleInfo::containedTypeHint_`: This is a type hint that applies only to the submodule associated with the `ModuleInfo` instance only when it is accessed as a submodule of the module associated with the `ConcreteModuleTypeBuilder` instance. That's why the type hint is threaded through to `ModuleValue` instead of being accessed from the `ModuleValue::concreteType_`; it does not exist on the concrete type of the submodule.
```
        class Mod(torch.nn.Module):
            # Calls addModule on the ConcreteModuleTypeBuilder associated 
            # with ModDict and passes Dict[str, ModuleInterface] as the hint for the 
            # submodule named d
            __annotations__ = {"d": Dict[str, ModuleInterface]}

            def __init__(self):
                super().__init__()
                self.d = torch.nn.ModuleDict({"module": ImplementsInterface()})

            def forward(self, x: torch.Tensor, key: str) -> Any:
                return self.d[key].forward(x)
```

I am reluctant to unify them because when/if we fix module aliasing and allow the same script module to be referred to in multiple places, always setting the contained type hint on the concrete type of the module it refers to (option 1 above) will no longer make sense.

Reply:
Disregard this, I think type sharing is not a concern here. As you pointed out below, setting a type hint on an instance of `ConcreteModuleTypeBuilder` will cause `equals` to return `false` when `ConcreteTypeStore.get_or_create_concrete_type` tries to find an existing type that matches the concrete type for the module currently being compiled.

Reply:
Disregard this, I think you're right in saying that type sharing is not a concern here because setting a type hint on a concrete type will make it compare unequal to any concrete type that would be equal to it in the absence of that type hint.